library/alloc/src/boxed.rs

```diff
@@ -1016,6 +1016,165 @@ impl<T: ?Sized, A: Allocator> Box<T, A> {
+    /// # Errors
+    ///
+    /// This function fails if `T` and `U` do not have the same [memory layout](Layout).
+    ///
+    /// # Examples
+    ///
+    /// Reuse a boxed slice:
+    ///
+    /// ```
+    /// #![feature(box_reuse)]
+    ///
+    /// let mut numbers: Box<[i32]> = Box::new([1, 2, 3, 4, 5]);
+    /// Box::reuse(&mut numbers, [6, 7, 8, 9, 10]).unwrap();
+    /// assert_eq!(&*numbers, &[6, 7, 8, 9, 10]);
+    /// ```
+    ///
+    /// Reuse a box to store an opaque type:
+    ///
+    /// ```
+    /// #![feature(box_reuse)]
+    ///
+    /// fn iterate() -> impl Iterator<Item = i32> {
+    ///     0..3
+    /// }
+    ///
+    /// let mut iterator: Box<dyn Iterator<Item = i32>> = Box::new(iterate());
+    /// assert_eq!(iterator.by_ref().collect::<Vec<_>>(), [0, 1, 2]);
+    ///
+    /// assert!(Box::reuse(&mut iterator, iterate()).is_ok());
+    /// assert_eq!(iterator.collect::<Vec<_>>(), [0, 1, 2]);
+    /// ```
+    #[unstable(feature = "box_reuse", issue = "none")]
+    pub fn reuse<U>(b: &mut Box<T, A>, val: U) -> Result<(), U>
+    where
+        U: Unsize<T>,
+    {
+        if Layout::for_value(&**b) != Layout::new::<U>() {
+            return Err(val);
+        }
+
+        struct Guard<'a, T: ?Sized, A: Allocator, U: Unsize<T>> {
+            b: &'a mut Box<T, A>,
+            val: mem::ManuallyDrop<U>,
+        }
+        impl<T: ?Sized, A: Allocator, U: Unsize<T>> Drop for Guard<'_, T, A, U> {
+            fn drop(&mut self) {
+                let ptr: Unique<U> = self.b.0.cast();
+                unsafe { ptr::write(ptr.as_ptr(), mem::ManuallyDrop::take(&mut self.val)) };
+                self.b.0 = ptr;
+            }
+        }
+
+        // Even if dropping the old value panics, this guard ensures that the new value will always
+        // be written.
+        let guard = Guard { b, val: mem::ManuallyDrop::new(val) };
+
+        unsafe { ptr::drop_in_place(&mut **guard.b) };
+
+        Ok(())
+    }
+
+    /// Sets the value inside the box, reusing the existing storage if possible.
+    ///
+    /// # Examples
+    ///
+    /// Reuse a boxed slice:
+    ///
+    /// ```
+    /// #![feature(box_reuse)]
+    ///
+    /// let mut numbers: Box<[i32]> = Box::new([1, 2, 3]);
+    ///
+    /// // This will not reallocate.
+    /// Box::set(&mut numbers, [4, 5, 6]);
+    /// assert_eq!(&*numbers, &[4, 5, 6]);
+    ///
+    /// // This will reallocate.
+    /// Box::set(&mut numbers, [4, 5, 6, 7]);
+    /// assert_eq!(&*numbers, &[4, 5, 6, 7]);
+    /// ```
+    #[unstable(feature = "box_reuse", issue = "none")]
+    #[cfg(not(no_global_oom_handling))]
+    pub fn set<U>(b: &mut Box<T, A>, val: U)
+    where
+        U: Unsize<T>,
+    {
+        let layout = Layout::new::<U>();
+        if Box::try_set(b, val).is_err() {
+            handle_alloc_error(layout);
+        }
+    }
+
+    /// Sets the value inside the box, reusing the existing storage if possible and returning an
+    /// error if the allocation fails.
```

Comment:
"reusing the existing storage if possible" seems to imply that this function will return an error if reusing the storage is not possible, however "returning an error if the allocation fails" seems to imply that this function will try to allocate a new `Box` if reusing the existing storage is not possible. 

Maybe this could be reflected in the function names, for example `try_set` could be renamed to `try_set_or_alloc`/`try_set_or_realloc`/`try_set_or_alloc_new` and `reuse` could be renamed to `try_set`

Reply:
I think using `or` could be a good idea, but I feel like `alloc` is too low-level to include in there. `replace` might be better. I would also like to keep the pattern of `X` and `try_X` being for infallible and fallible allocations respectively, having `reuse` be `try_set` breaks that. Changing `[try_]set` to be `[try_]reuse_or_replace` is an option, although it's quite wordy.